stores/cloudflare-d1/src/storage/index.ts

```diff
@@ -652,108 +724,203 @@ export class D1Store extends MastraStorage {
       if (format === `v2`) return list.get.all.v2();
       return list.get.all.v1();
+    };
+
+    try {
+      if (page !== undefined) {
+        // Paginated path
+        const perPage = perPageInput ?? 40;
+        const currentOffset = page * perPage;
+
+        const conditions: string[] = [`thread_id = ?`];
+        const queryParams: SqlParam[] = [threadId];
+
+        if (fromDate) {
+          conditions.push(`createdAt >= ?`);
+          queryParams.push(this.serializeDate(fromDate));
+        }
+        if (toDate) {
+          conditions.push(`createdAt <= ?`);
+          queryParams.push(this.serializeDate(toDate));
+        }
+        const whereClause = conditions.join(' AND ');
+
+        const countQueryBuilder = createSqlBuilder()
+          .count()
+          .from(fullTableName)
+          .where(whereClause, ...queryParams);
+        const { sql: countSql, params: countParams } = countQueryBuilder.build();
+        const countResult = (await this.executeQuery({ sql: countSql, params: countParams, first: true })) as {
+          count: number;
+        } | null;
+        const total = Number(countResult?.count || 0);
+
+        if (total === 0) {
+          return {
+            messages: [],
+            total: 0,
+            page,
+            perPage,
+            hasMore: false,
+          };
+        }
+
+        const dataQueryBuilder = createSqlBuilder()
+          .select(['id', 'content', 'role', 'type', 'createdAt', 'thread_id AS threadId'])
+          .from(fullTableName)
+          .where(whereClause, ...queryParams)
+          .orderBy('createdAt', 'DESC')
+          .limit(perPage)
+          .offset(currentOffset);
+        const { sql: dataSql, params: dataParams } = dataQueryBuilder.build();
+        const rows = await this.executeQuery({ sql: dataSql, params: dataParams });
+        const allMessages = Array.isArray(rows) ? [...rows] : [];
+
+        const messages = formatAndDedupeMessages(allMessages);
+
+        return {
+          messages: messages,
+          total,
+          page,
+          perPage,
+          hasMore: currentOffset + allMessages.length < total,
+        };
+      } else {
+        // Non-paginated path
+        const limit = typeof selectBy?.last === `number` ? selectBy.last : 40;
+        const include = selectBy?.include || [];
+        let messages: any[] = []; // Keep as any[] for intermediate results
+
+        if (include.length) {
+          const prevMax = Math.max(0, ...include.map(i => i.withPreviousMessages || 0));
+          const nextMax = Math.max(0, ...include.map(i => i.withNextMessages || 0));
+          const includeIds = include.map(i => i.id);
+
+          const sql = `
+            WITH ordered_messages AS (
+              SELECT
+                * ,
+                ROW_NUMBER() OVER (ORDER BY createdAt DESC) AS row_num
+              FROM ${fullTableName}
+              WHERE thread_id = ?
+            ),
+            target_message_details AS (
```

Comment:
This new code is looking pretty rough to read. Makes me think we should keep the existing methods as is and add new methods like `getMessagesPaginated` where it just wraps `getMessages` and `getMessages` takes in a new input arg for offset.

I think this is also implemented wrong for the public API. Instead of adding new args like fromDate/toDate/etc those should be part of the existing `selectBy` object which is a kind of query settings object.
Do we need fromDate/toDate though? If we do they should be something like `dateRange: { start: x, end: y }`